### ISSUE TITLE
Modern Private Field Enforcement.

### ISSUE THREAD

**Issue by jimmywarting** [gajus/eslint-plugin-jsdoc#1633]
## Motivation

As ECMAScript private class fields (`#field`) become the standard for encapsulation, JSDoc usage needs to evolve. Currently, there is a disconnect between documentation and implementation:

1. **Enforcement:** `@private` is often used on public fields, creating a false sense of security.
2. **Redundancy:** Once `#` is used, the `@private` tag becomes "noise" because the language itself enforces the visibility.

Combining these rule logic ensures that the `#` symbol is the **single source of truth** for privacy in a codebase.

---

## Current behavior

The plugin currently allows both "soft" privacy (using `@private` on a public field) and "redundant" privacy (using `@private` on a `#` field).

```javascript
class Example {
  /** @private */
  softPrivate = 1; // Allowed, but insecure.

  /** @private */
  #hardPrivate = 2; // Allowed, but redundant.
}

```

---

## Desired behavior

I propose a new rule, e.g., `jsdoc/require-hash-with-private-tag`, with two primary configurations to handle these scenarios separately:

### 1. Requirement: `@private` demands `#` syntax

If a field is documented as `@private`, the rule should require the field to use the `#` prefix. This prevents developers from accidentally exposing "private" data to the runtime API.

**Failing Code:**

```javascript
/** @private */
count = 0; // Error: Use #count instead.

```

### 2. Requirement: `#` syntax forbids `@private` tag

If a field uses the `#` prefix, the rule should flag the `@private` tag as redundant and suggest removal.

**Failing Code:**

```javascript
/**
 * @private
 * Internal counter.
 */
#count = 0; // Error: @private is redundant with # syntax.

```

**Final Desired State:**

```javascript
/** Internal counter. */
#count = 0; 

```

```js
export default {
	'require-hash-with-private-tag': {
		meta: {
			type: 'problem',
			docs: {
				description: 'Require # symbol on class fields that use @private JSDoc tag',
				category: 'Best Practices'
			},
			messages: {
				requireHash: 'Class fields with @private tag must use the # symbol to be truly private'
			},
			schema: []
		},
		create: (context) => {
			const { sourceCode } = context
			
			return {
				PropertyDefinition (node) {
					// Check if the property does NOT use the private # symbol
					if (!node.key || node.key.type !== 'PrivateIdentifier') {
						// Get JSDoc comment for this node
						const comments = sourceCode.getCommentsBefore(node)
						
						for (const comment of comments) {
							if (comment.type === 'Block' && comment.value.includes('*')) {
								// Check if it contains @private tag (as a JSDoc tag, not just the word "private")
								// Match @private at the beginning of a line with optional whitespace and asterisk
								if (/^\s*\*\s*@private\b/m.test(comment.value)) {
									context.report({
										node,
										messageId: 'requireHash'
									})
								}
							}
						}
					}
				}
			}
		}
	},
	'no-private-tag-with-hash': {
		meta: {
			type: 'problem',
			docs: {
				description: 'Disallow @private JSDoc tag on class fields that use the # symbol',
				category: 'Best Practices'
			},
			messages: {
				noPrivateWithHash: 'Remove @private tag - the # symbol already indicates this field is private'
			},
			schema: []
		},
		create: (context) => {
			const { sourceCode } = context
			
			return {
				PropertyDefinition (node) {
					// Check if the property uses the private # symbol
					if (node.key && node.key.type === 'PrivateIdentifier') {
						// Get JSDoc comment for this node
						const comments = sourceCode.getCommentsBefore(node)
						
						for (const comment of comments) {
							if (comment.type === 'Block' && comment.value.includes('*')) {
								// Check if it contains @private tag (as a JSDoc tag, not just the word "private")
								// Match @private at the beginning of a line with optional whitespace and asterisk
								if (/^\s*\*\s*@private\b/m.test(comment.value)) {
									context.report({
										node,
										messageId: 'noPrivateWithHash'
									})
								}
							}
						}
					}
				}
			}
		}
	}
}
```

**Comment by jimmywarting** [comment]
I guess having this two rules would help with auto fixes such that it

1. Must first add the `#` symbol to the code
2.  Then remove the `@private` tag

(rather than having one combined rule - also allows it to be more flexible)

**Comment by brettz9** [comment]
> Also now that ppl are using ES modules more and more, `@public` also becomes less relevant now that we can properly export something and "declare" what should be public or not. otherwise it's tied to it's own file scope. and being private in that manner.
> 
> so `@package` is being less and less relevant also.

Since one can entirely block tags in `check-tag-names` already, I think that might be a better place for it.

Btw, I think we probably need to allow `@protected` and `@access protected`.
